src/system-shutdown.c

```diff
@@ -1,153 +1,235 @@
+{
+	int fd = open(src, O_RDONLY);
+	if (fd < 0) {
+		kmsg("* unable to open loop device %s: %s", src,
+		     strerror(errno));
+	} else {
+		if (ioctl(fd, LOOP_CLR_FD) < 0) {
+			kmsg("* unable to disassociate loop device %ss: %s",
+			     src, strerror(errno));
+		}
+		close(fd);
+	}
+}
 
 // tries to umount all (well, most) things. Returns whether in the last pass it
 // no longer found writable.
-static bool umount_all() {
-        bool did_umount = true;
-        bool had_writable = false;
-
-        for (int i=0; i<10 && did_umount; i++) {
-                struct mountinfo *mounts = parse_mountinfo(NULL);
-                if (!mounts) {
-                        // oh dear
-                        die("unable to get mount info; giving up");
-                }
-                struct mountinfo_entry *cur = first_mountinfo_entry(mounts);
-
-                had_writable = false;
-                did_umount = false;
-                while (cur) {
-                        const char* mount_dir = mountinfo_entry_mount_dir(cur);
-                        const char* mount_src = mountinfo_entry_mount_source(cur);
-                        cur = next_mountinfo_entry(cur);
-
-                        if (strcmp("/", mount_dir) == 0) {
-                                continue;
-                        }
-
-                        if (strcmp("/dev", mount_dir) == 0) {
-                                continue;
-                        }
-
-                        if (strcmp("/proc", mount_dir) == 0) {
-                                continue;
-                        }
-
-                        if (strstr(mount_dir, "writable")) {
-                                had_writable = true;
-                        }
-
-                        if (umount(mount_dir) == 0) {
-                                int fd = open(mount_src, O_RDONLY);
-                                if (fd >=0) {
-                                        ioctl(fd, LOOP_CLR_FD);
-                                        close(fd);
-                                }
-
-                                did_umount = true;
-                        }
-                }
-                cleanup_mountinfo(&mounts);
-        }
-
-        return !had_writable;
+static bool umount_all()
+{
+	bool did_umount = true;
+	bool had_writable = false;
+
+	for (int i = 0; i < 10 && did_umount; i++) {
+		struct mountinfo *mounts = parse_mountinfo(NULL);
+		if (!mounts) {
+			// oh dear
+			die("unable to get mount info; giving up");
+		}
+		struct mountinfo_entry *cur = first_mountinfo_entry(mounts);
+
+		had_writable = false;
+		did_umount = false;
+		while (cur) {
+			const char *dir = mountinfo_entry_mount_dir(cur);
+			const char *src = mountinfo_entry_mount_source(cur);
+			unsigned major = mountinfo_entry_dev_major(cur);
+
+			cur = next_mountinfo_entry(cur);
+
+			if (streq("/", dir)) {
+				continue;
+			}
+
+			if (streq("/dev", dir)) {
+				continue;
+			}
+
+			if (streq("/proc", dir)) {
```

Comment:
I assume it is because of /proc/self/mountinfo but it would be good to document why this is not unmounted.